File: APIs/PanelesSolaresAPI/core/routers/producto_router.py
from fastapi import APIRouter, Path
import requests
import json

from assets.util import url_base

from core.models.todos_model import producto_actualizar, producto_crear

url_productos = f'{url_base}/productos/'

router_producto = APIRouter(
    prefix="/productos",
    tags=["productos"]
)

@router_producto.get('/todos')
def listar_productos():
    payload = ""
    headers = {}

    response = requests.request("GET", url_productos, headers=headers, data=payload)
    
    return response.json()

@router_producto.get('/uno/{producto_id}')
def buscar_producto_por_id(producto_id: int = Path(title="El ID debe ser un número positivo", gt=0)):
    
    url = f'{url_productos}/{producto_id}'
    payload = ""
    headers = {}
    response = requests.request("GET", url, headers=headers, data=payload)
    
    return response.json()

@router_producto.post('/crear')
def crear_producto(producto: producto_crear):
    print(producto)
    payload = json.dumps({
        "nombre": producto.nombre,
        "referencia": producto.referencia,
        "descripcion": producto.descripcion,
        "marca": producto.marca,
        "cantidad_minima": producto.cantidad_minima,
        "cantidad_actual": producto.cantidad_actual,
        "precio_compra": producto.precio_compra,
        "precio_venta": producto.precio_venta
    })
    headers = {
    'Content-Type': 'application/json'
    }

    response = requests.request("POST", url_productos, headers=headers, data=payload)
    return response.json()

@router_producto.delete('/eliminar/{producto_id}')
def eliminar_producto_por_id(producto_id: int):
    
    url = f'{url_productos}/{producto_id}'

    payload = {}
    headers = {}

    response = requests.request("DELETE", url, headers=headers, data=payload)
    return response.json()

@router_producto.put('/actualizar')
def actualizar_producto_por_id(producto: producto_actualizar):
    url = f'{url_productos}/{producto.id}'

    payload = json.dumps({
        "nombre": producto.nombre,
        "referencia": producto.referencia,
        "descripcion": producto.descripcion,
        "marca": producto.marca,
        "cantidad_minima": producto.cantidad_minima,
        "cantidad_actual": producto.cantidad_actual,
        "precio_compra": producto.precio_compra,
        "precio_venta": producto.precio_venta
    })
    headers = {
    'Content-Type': 'application/json'
    }

    response = requests.request("PUT", url, headers=headers, data=payload)
    return response.json()

File: APIs/APINodeJS/src/routes/producto.router.ts
import express, { Request, Response } from 'express';
import { producto, productoAttributes } from '../models/init-models.model'; // Asegúrate de que la ruta del modelo sea correcta
import { Sequelize } from 'sequelize';

const router = express.Router();

// Ruta para crear un nuevo producto
router.post('/', async (req: Request, res: Response) => {
  try {
    const nuevoProducto = await producto.create(req.body as productoAttributes);
    return res.json(nuevoProducto);
  } catch (error) {
    console.error('Error al crear un producto:', error);
    return res.status(500).json({ error: 'Error al crear un producto' });
  }
});

// Ruta para obtener todos los productos
router.get('/', async (req: Request, res: Response) => {
  try {
    const productos = await producto.findAll();
    return res.json(productos);
  } catch (error) {
    console.error('Error al obtener productos:', error);
    return res.status(500).json({ error: 'Error al obtener productos' });
  }
});

// Ruta para obtener un producto por ID
router.get('/:id', async (req: Request, res: Response) => {
  const { id } = req.params;
  try {
    const productoEncontrado = await producto.findByPk(Number(id));
    if (!productoEncontrado) {
      return res.status(404).json({ error: 'Producto no encontrado' });
    }
    return res.json(productoEncontrado);
  } catch (error) {
    console.error('Error al obtener producto por ID:', error);
    return res.status(500).json({ error: 'Error al obtener producto por ID' });
  }
});

// Ruta para actualizar un producto por ID
router.put('/:id', async (req: Request, res: Response) => {
  const { id } = req.params;
  try {
    const productoActualizado = await producto.findByPk(Number(id));
    if (!productoActualizado) {
      return res.status(404).json({ error: 'Producto no encontrado' });
    }
    await productoActualizado.update(req.body as productoAttributes);
    return res.json(productoActualizado);
  } catch (error) {
    console.error('Error al actualizar producto por ID:', error);
    return res.status(500).json({ error: 'Error al actualizar producto por ID' });
  }
});

// Ruta para eliminar un producto por ID
router.delete('/:id', async (req: Request, res: Response) => {
  const { id } = req.params;
  try {
    const productoEliminado = await producto.findByPk(Number(id));
    if (!productoEliminado) {
      return res.status(404).json({ error: 'Producto no encontrado' });
    }
    await productoEliminado.destroy();
    return res.json({ message: 'Producto eliminado con éxito' });
  } catch (error) {
    console.error('Error al eliminar producto por ID:', error);
    return res.status(500).json({ error: 'Error al eliminar producto por ID' });
  }
});

router.post('/producto', async (req: Request, res: Response) => {
  try {
    const { producto_id } = req.body;

    const query = `
      SELECT 
        p.id AS producto_id,
        p.nombre AS producto_nombre,
        p.referencia,
        p.descripcion,
        p.marca,
        p.cantidad_minima,
        p.cantidad_actual,
        p.precio_compra,
        p.precio_venta,
        p.fecha_creacion AS producto_fecha_creacion,
        p.fecha_modificacion AS producto_fecha_modificacion,
        i.nombre AS imagen_nombre,
        i.url AS imagen_url
      FROM producto p
      LEFT JOIN imagen i ON p.id = i.producto_id
      WHERE p.id = '${producto_id}';
    `;

    // Obtener instancia de Sequelize
    const sequelize: Sequelize | undefined = producto.sequelize;

    if (!sequelize) {
      throw new Error('No es posible obtener la referencia al objeto Sequelize');
    }

    // Ejecutar la consulta
    const [result]: any = await sequelize.query(query);

    // Retornar la información del producto
    return res.json(result);
  } catch (error) {
    console.error('Error al obtener información del producto:', error);
    return res.status(500).json({ error: 'Error al obtener información del producto' });
  }
});

export default router;
